clash-cores/src/Clash/Cores/Xilinx/DcFifo/Implementation.hs

```diff
@@ -0,0 +1,282 @@
+-- behaviors are functionally correct and equivalent.
+--
+-- If 'dcReadDataCount', 'dcUnderflow', 'dcOverflow', or 'dcWriteDataCount' are disabled, the relevant signals
+-- will be 'deepErrorX'.
+dcFifo ::
+  forall depth a n write read .
+  ( KnownDomain write
+  , KnownDomain read
+
+  , KnownNat n
+  , BitPack a
+  , n ~ BitSize a
+  , NFDataX a
+
+  , KnownNat depth
+  -- Number of elements should be between [2**4, 2**17] ~ [16, 131072].
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  -- | @wr_clk@
+  Clock write ->
+  -- | @rd_clk@
+  Clock read ->
+  -- | Reset for both read and write side
+  Reset read ->
+
+  -- | Write data, @din@ and @wr_en@
+  Signal write (Maybe a) ->
+  -- | Read enable @rd_en@
+  Signal read Bool ->
+  FifoOut read write depth a
+dcFifo DcConfig{..} wClk rClk rst writeData rEnable =
+  let
+    (wFull, wOver, wCnt, rEmpty, rUnder, rCnt, rData) =
+      go initState rstSignalR rEnable rstSignalW writeData
+  in FifoOut
+      wFull
+      (if dcOverflow then wOver else deepErrorX "Overflow disabled")
+      (if dcWriteDataCount then wCnt else deepErrorX "Write data count disabled")
+      rEmpty
+      (if dcUnderflow then rUnder else deepErrorX "Underflow disabled")
+      (if dcReadDataCount then rCnt else deepErrorX "Read data count disabled")
+      (deepErrorX "No sample" :- rData)
+
+ where
+  rstSignalR = unsafeToHighPolarity rst
+  rstSignalW = unsafeSynchronizer rClk wClk $ unsafeToHighPolarity rst
+
+  -- reified depth
+  maxDepth = snatToNum @Int (powSNat (SNat @2) (SNat @depth))
+
+  go ::
+    FifoState a ->
+    Signal read Bool -> -- reset
+    Signal read Bool -> -- read enabled
+    Signal write Bool -> -- reset
+    Signal write (Maybe a) -> -- write data
+    ( Signal write Full
+    , Signal write Bool
+    , Signal write (DataCount depth)
+
+    , Signal read Empty
+    , Signal read Bool
+    , Signal read (DataCount depth)
+    , Signal read a
+    )
+  go st@(FifoState _ rt) rstR rEna rstW =
+    if rt < tWr
+      then goRead st rstR rEna rstW
+      else goWrite st rstR rEna rstW
+
+  goWrite (FifoState _ rt) rstR rEna (True :- rstWNext) (_ :- wData) =
+      (False :- preFull, False :- preOver, 0 :- preWCnt, fifoEmpty, under, rCnt, rData)
+    where
+      (preFull, preOver, preWCnt, fifoEmpty, under, rCnt, rData) =
+        go (FifoState mempty (rt-tWr)) rstR rEna rstWNext wData
+
+  goWrite (FifoState q rt) rstR rEna (_ :- rstW) (wDat :- wDats1) =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      (preFull, preOver, preWCnt, fifoEmpty, under, rCnt, rData) =
+        go (FifoState q' (rt-tWr)) rstR rEna rstW wDats1
+
+      wCnt = sDepth q :- preWCnt
+      full = (Seq.length q == maxDepth) :- preFull
+      (q', over) =
+        if Seq.length q < maxDepth
+          then (case wDat of { Just x -> x Seq.<| q ; _ -> q }, False :- preOver)
+          else (q, True :- preOver)
```

Comment:
> please also verify whether overflow is asserted combinatorially or through a register!

According to the manual for the FIFO generator, the overflow signal has 0 latency after a write operation (table 3-19 for us). So I guess that means it's asserted combinatorially?

> which of the output flags are combinatorially determined?

Looks like `wr_data_count` has a latency of 1 so I guess that's through a register. But `full` and `empty` have a latency of zero 🤔 

If I try to make `full` and `empty` instantaneous (by setting `fifoEmpty = (Seq.length q' == 0) :- preEmpty`), however, this causes an infinite loop in our Haskell code.

Reply:
> If I try to make full and empty instantaneous (by setting fifoEmpty = (Seq.length q' == 0) :- preEmpty), however, this causes an infinite loop in our Haskell code.

Most likely that's a bug in how we use the `empty` signal in the test bench; if `empty` is combinatorially determined the test bench can't also combinatorially determine `rd_en` based on `empty` because that is a combinatorial loop. We need to change the test bench so it registers its `rd_en`.

Reply:
Right, ok modifying the test bench to use a register works! Which I guess makes sense. 

(Perhaps I am not understanding but I think something like 06b3a339c9a59800fb859814bf5ba3900d539f8b is necessary for it to have latency zero?)

Reply:
I don't think that would be a correct fix (you'd just be misinterpreting the signal), but also, as I said, I don't believe they _are_ combinatorial. For one thing, you'd expect the Vivado simulation to hang on a combinatorial loop as well. Other than that, if I look at figure 3-2 and 3-3 in [pg057-fifo-generator](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator), those signals seem pretty registered to me. We read/write three consecutive items and then it's full/empty. Pretty much everything is registered, all the outputs for the third read/write are in cycle 4.